storage/remote/queue_manager_test.go

```
@@ -164,6 +165,186 @@ func TestSampleDelivery(t *testing.T) {
+
+	series, samples := createTestClassicHistogram(10)
+
+	for _, tc := range []struct {
+		name                   string
+		queueConfig            config.QueueConfig
+		expRequestSamples      [][]record.RefSample
+		allowUnOrderedRequests bool
+	}{
+		{
+			name: "OneShardDefaultBatch",
+			queueConfig: func() config.QueueConfig {
+				cfg := testDefaultQueueConfig()
+				cfg.MaxShards = 1
+				cfg.MinShards = 1
+				return cfg
+			}(),
+			expRequestSamples: [][]record.RefSample{samples},
+		},
+		{
+			name: "OneShardLimitedBatch",
+			queueConfig: func() config.QueueConfig {
+				cfg := testDefaultQueueConfig()
+				cfg.MaxShards = 1
+				cfg.MinShards = 1
+				cfg.MaxSamplesPerSend = 5
+				return cfg
+			}(),
+			expRequestSamples: [][]record.RefSample{
+				samples[:5], samples[5:10], samples[10:],
+			},
+		},
+		{
+			name: "TwoShardsLimitedBatch",
+			queueConfig: func() config.QueueConfig {
+				cfg := testDefaultQueueConfig()
+				cfg.MaxShards = 2
+				cfg.MinShards = 2
+				return cfg
+			}(),
+			expRequestSamples: [][]record.RefSample{
+				{samples[0], samples[2], samples[4], samples[6], samples[8], samples[10]},
+				{samples[1], samples[3], samples[5], samples[7], samples[9], samples[11]},
+			},
+			allowUnOrderedRequests: true, // No order guarantee for requests from shards.
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			c := newPerRequestWriteClient(tc.allowUnOrderedRequests)
+
+			for _, s := range tc.expRequestSamples {
+				c.expectRequestSamples(s, series)
+			}
+
+			dir := t.TempDir()
+			mcfg := config.DefaultMetadataConfig
+
+			metrics := newQueueManagerMetrics(nil, "", "")
+			m := NewQueueManager(metrics, nil, nil, nil, dir, newEWMARate(ewmaWeight, shardUpdateDuration), tc.queueConfig, mcfg, labels.EmptyLabels(), nil, c, defaultFlushDeadline, newPool(), newHighestTimestampMetric(), nil, false, false, MinStrings)
+			m.StoreSeries(series, 0)
+
+			m.Start()
+			defer m.Stop()
+
+			m.Append(samples)
 			c.waitForExpectedData(t)
```

Since que queue manager waits for pending samples to be sent when stop, can't we do:

```go
m.Start()
m.Append(samples)
m.Stop()
c.assertExpectedData(t)
```
So the test fails fast instead of timing out. 

Ofc I might be missing something 🙂

---

Will check, not sure now, my understanding was that there is some async behavior